kge/job/train.py

```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
+            - label coordinates (position of ones in a n x num_entities or n x num_relations tensor for so respectively)
+            - query_type (vector of size n, 2 if corresponding example_index is so, 1 if sp, 0 if po)
             - triples (all true triples in the batch: needed for weighted penalties only)
 
             """
+
             # count how many labels we have
             num_ones = 0
             for example_index in batch:
-                if example_index < num_sp:
-                    num_ones += self.train_sp_offsets[example_index + 1]
-                    num_ones -= self.train_sp_offsets[example_index]
-                else:
-                    example_index -= num_sp
-                    num_ones += self.train_po_offsets[example_index + 1]
-                    num_ones -= self.train_po_offsets[example_index]
+                prev_end = 0
+                for query, end in self.query_end_index.items():
+                    if example_index < end:
+                        example_index -= prev_end
+                        num_ones += self.offsets[query][example_index + 1]
+                        num_ones -= self.offsets[query][example_index]
+                        break
+                    prev_end = end
 
             # now create the results
-            sp_po_batch = torch.zeros([len(batch), 2], dtype=torch.long)
-            is_sp = torch.zeros([len(batch)], dtype=torch.long)
+            key_batch = torch.zeros([len(batch), 2], dtype=torch.long)
+            query_type = torch.zeros([len(batch)], dtype=torch.long)
             label_coords = torch.zeros([num_ones, 2], dtype=torch.int)
             current_index = 0
             triples = torch.zeros([num_ones, 3], dtype=torch.long)
             for batch_index, example_index in enumerate(batch):
-                is_sp[batch_index] = 1 if example_index < num_sp else 0
-                if is_sp[batch_index]:
-                    keys = self.train_sp_keys
-                    offsets = self.train_sp_offsets
-                    values = self.train_sp_values
-                    sp_po_col_1, sp_po_col_2, o_s_col = S, P, O
-                else:
-                    example_index -= num_sp
-                    keys = self.train_po_keys
-                    offsets = self.train_po_offsets
-                    values = self.train_po_values
-                    o_s_col, sp_po_col_1, sp_po_col_2 = S, P, O
-
-                sp_po_batch[batch_index,] = keys[example_index]
+                prev_end = 0
+                for query_index, (query, end) in enumerate(self.query_end_index.items()):
+                    if example_index < end:
+                        example_index -= prev_end
+                        query_type[batch_index] = query_index
+                        keys = self.keys[query]
+                        offsets = self.offsets[query]
+                        values = self.values[query]
+                        if query_index == 0:
```

Comment:
Woudln't it be more natural if query_index refers to the position of the element being predicted? (`0=_po, 1=s_o, 2=sp_`)?